### PR TITLE
Flexible billing primitives

### PR DESCRIPTION
**Pull request by alexander-stripe**
PR ported from internal internal GHE.

Adds flexible billing primitives to the SDK for beta functionality.

### CODE REVIEW

File: plan_test.go [hunk 1]
```diff
@@ -44,22 +44,34 @@ func TestPlanParams_AppendTo(t *testing.T) {
...
+		{"product", &PlanParams{ProductID: &productID}, "prod_123abc"},
 		{"trial_period_days", &PlanParams{TrialPeriod: 123}, strconv.FormatUint(123, 10)},
+		{"tiers_mode", &PlanParams{TiersMode: "volume"}, "volume"},
+		{"usage_type", &PlanParams{UsageType: "metered"}, "metered"},
+		{"transform_usage[bucket_size]", &PlanParams{TransformUsage: &PlanTransformUsageParams{DivideBy: 123, Round: "round_up"}}, strconv.FormatUint(123, 10)},
+		{"transform_usage[round]", &PlanParams{TransformUsage: &PlanTransformUsageParams{DivideBy: 123, Round: "round_up"}}, "round_up"},
+		{"tiers[0][amount]", &PlanParams{Tiers: tiers}, strconv.FormatUint(123, 10)},
+		{"tiers[0][up_to]", &PlanParams{Tiers: tiers}, strconv.FormatUint(321, 10)},
+		{"tiers[1][amount]", &PlanParams{Tiers: tiers}, strconv.FormatUint(123, 10)},
+		{"tiers[1][up_to]", &PlanParams{Tiers: tiers}, "inf"},
```
Comment (brandur-stripe) [inline plan_test.go]: Sorry for the nit, but would you mind just alphabetizing these test cases by their leading string?

File: usagerecord/client_test.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+	_ "github.com/stripe/stripe-go/testing"
+)
+
+func TestUsageRecordNew(t *testing.T) {
+	now := uint64(time.Now().Unix())
+	plan, err := New(&stripe.UsageRecordParams{
+		Quantity:         123,
+		Timestamp:        now,
+		Action:           stripe.UsageRecordParamsActionIncrement,
+		SubscriptionItem: "si_abc123",
```
Comment (brandur-stripe) [inline usagerecord/client_test.go]: Just per convention, mind just changing this to `si_123`?

**Comment by brandur-stripe** [comment]
Added two minor asks, but otherwise LGTM.

Would you also mind just rebasing this into one commit with a good title/description? We've been trying to improve our commit hygiene more recently. Thanks!

**Comment by alexander-stripe** [comment]
Thanks for taking a look! Done.

**Comment by brandur-stripe** [comment]
Nice! Thanks for the quick fixes.

**Comment by brandur-stripe** [comment]
Released as 30.3.0.

**Pull request merged**